src/main/java/com/iota/iri/utils/thread/BoundedScheduledExecutorService.java

```diff
@@ -0,0 +1,931 @@
+     * <br />
+     * In addition to delegating the method call to the internal {@link ScheduledExecutorService}, we call the cancel
+     * logic for recurring tasks because shutdown prevents them from firing again. If these "cancelled" jobs are
+     * scheduled for execution (and not running right now), we also call their
+     * {@link #onCompleteTask(TaskDetails, Throwable)} callback to "report" that hey have finished (otherwise this will
+     * be fired inside the wrapped task).<br />
+     */
+    @Override
+    public void shutdown() {
+        delegate.shutdown();
+
+        for (TaskDetails currentTask : scheduledTasks) {
+            if (currentTask.getInterval() != null) {
+                onCancelTask(currentTask);
+
+                if (currentTask.getScheduledForExecution().compareAndSet(true, false)) {
+                    onCompleteTask(currentTask, null);
+                }
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * <br />
+     * Before delegating the method call to the internal {@link ScheduledExecutorService}, we call the
+     * {@link #onCancelTask(TaskDetails)} callback for all scheduled tasks and fire the
+     * {@link #onCompleteTask(TaskDetails, Throwable)} callback for all tasks that are not being executed right now
+     * (otherwise this will be fired inside the wrapped task).<br />
+     */
+    @Override
+    public List<Runnable> shutdownNow() {
+        if (!delegate.isShutdown()) {
+            for (TaskDetails currentTask : scheduledTasks) {
+                onCancelTask(currentTask);
+
+                if (currentTask.getScheduledForExecution().compareAndSet(true, false)) {
+                    onCompleteTask(currentTask, null);
+                }
+            }
+        }
+
+        return delegate.shutdownNow();
+    }
+
+    @Override
+    public boolean isShutdown() {
+        return delegate.isShutdown();
+    }
+
+    @Override
+    public boolean isTerminated() {
+        return delegate.isTerminated();
+    }
+
+    @Override
+    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+        return delegate.awaitTermination(timeout, unit);
+    }
+
+    //endregion ////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    //region PRIVATE UTILITY METHODS AND MEMBERS ///////////////////////////////////////////////////////////////////////
+
+    /**
+     * This interface is used to generically describe the lambda that is used to create the unwrapped future from the
+     * delegated {@link ScheduledExecutorService} by passing in the wrapped command.<br />
+     *
+     * @param <RESULT> the kind of future returned by this factory ({@link ScheduledFuture} vs {@link Future})
+     * @param <ARGUMENT> type of the wrapped command that is passed in ({@link Runnable} vs {@link Callable})
+     */
+    @FunctionalInterface
+    private interface FutureFactory<RESULT, ARGUMENT> {
```

Comment:
It is funny, I was so used to only using one letter in type params it never occured to me that we can write a word :laughing: 

Anyhows, :+1: on the clarity